zcash_keys/src/keys.rs

```diff
@@ -928,6 +928,351 @@ impl UnifiedFullViewingKey {
+                unified::Ivk::P2pkh(data) => legacy::ExternalIvk::deserialize(data)
+                    .map_err(|_| "Invalid transparent IVK in Unified IVK")
+                    .map(|tivk| {
+                        transparent = Some(tivk);
+                        None
+                    })
+                    .transpose(),
+                #[cfg(not(feature = "transparent-inputs"))]
+                unified::Ivk::P2pkh(data) => Some(Ok::<_, &'static str>((
+                    u32::from(unified::Typecode::P2pkh),
+                    data.to_vec(),
+                ))),
+                unified::Ivk::Unknown { typecode, data } => Some(Ok((*typecode, data.clone()))),
+            })
+            .collect::<Result<_, _>>()?;
+
+        Ok(Self {
+            #[cfg(feature = "transparent-inputs")]
+            transparent,
+            #[cfg(feature = "sapling")]
+            sapling,
+            #[cfg(feature = "orchard")]
+            orchard,
+            unknown,
+        })
+    }
+
+    /// Returns the string encoding of this `UnifiedFullViewingKey` for the given network.
+    pub fn encode<P: consensus::Parameters>(&self, params: &P) -> String {
+        let items = std::iter::empty().chain(self.unknown.iter().map(|(typecode, data)| {
+            unified::Ivk::Unknown {
+                typecode: *typecode,
+                data: data.clone(),
+            }
+        }));
+        #[cfg(feature = "orchard")]
+        let items = items.chain(
+            self.orchard
+                .as_ref()
+                .map(|ivk| ivk.to_bytes())
+                .map(unified::Ivk::Orchard),
+        );
+        #[cfg(feature = "sapling")]
+        let items = items.chain(
+            self.sapling
+                .as_ref()
+                .map(|divk| divk.to_bytes())
+                .map(unified::Ivk::Sapling),
+        );
+        #[cfg(feature = "transparent-inputs")]
+        let items = items.chain(
+            self.transparent
+                .as_ref()
+                .map(|tivk| tivk.serialize().try_into().unwrap())
+                .map(unified::Ivk::P2pkh),
+        );
+
+        let uivk = unified::Uivk::try_from_items(items.collect())
+            .expect("UnifiedFullViewingKey should only be constructed safely");
+        uivk.encode(&params.network_type())
+    }
+
+    #[cfg(feature = "transparent-inputs")]
+    fn transparent(&self) -> &Option<zcash_primitives::legacy::keys::ExternalIvk> {
+        &self.transparent
+    }
+
+    #[cfg(feature = "sapling")]
+    fn sapling(&self) -> &Option<::sapling::zip32::IncomingViewingKey> {
+        &self.sapling
+    }
+
+    #[cfg(feature = "orchard")]
+    fn orchard(&self) -> &Option<orchard::keys::IncomingViewingKey> {
+        &self.orchard
+    }
+
+    /// Attempts to derive the Unified Address for the given diversifier index and
+    /// receiver types.
+    ///
+    /// Returns `None` if the specified index does not produce a valid diversifier.
+    pub fn address(
```

Comment:
In order to prevent duplication, the `UnifiedFullViewingKey` methods that this is derived from should now delegate to this method.

Reply:
I absolutely agree, as per my PR description's checklist.